core/src/main/java/com/linecorp/armeria/internal/common/RequestContextUtil.java

```diff
@@ -48,13 +55,45 @@
     private static final Set<Thread> REPORTED_THREADS =
             Collections.newSetFromMap(new MapMaker().weakKeys().makeMap());
 
+    private static final ContextStorage contextStorage;
+
+    static {
+        final List<ContextStorageProvider> providers = Streams.stream(
+                ServiceLoader.load(ContextStorageProvider.class)).collect(toImmutableList());
+
+        if (providers.isEmpty()) {
+            contextStorage = new ContextStorageThreadLocal();
+        } else {
+            final ContextStorageProvider provider = providers.get(0);
+            if (providers.size() > 1) {
+                logger.warn("Found more than one {}. Only the first provider is used: {}, providers: {}",
+                            ContextStorageProvider.class.getSimpleName(), provider, providers);
+            }
+            ContextStorage temp;
+            try {
+                temp = provider.newContextStorage();
+            } catch (Throwable t) {
+                logger.warn("Failed to create context storage. provider: {}", provider, t);
```

Comment:
Think it's better to crash than fallback. As long as there is a flag to set the storage class (so could set it to default if really desired) it seems safer than a possibly ignored warning.